Lib/test/test_codecs.py

```diff
@@ -3415,5 +3415,30 @@ def test_rot13_func(self):
             'To be, or not to be, that is the question')
 
 
+class CodecNameNormalizationTest(unittest.TestCase):
+    """Test codec name normalization"""
+    def test_normalized_encoding(self):
+        FOUND = (1, 2, 3, 4)
+        NOT_FOUND = (None, None, None, None)
+        def search_function(encoding):
+            if encoding == "aaa_8":
+                return FOUND
+            else:
+                return NOT_FOUND
+
+        codecs.register(search_function)
+        self.assertEqual(FOUND, codecs.lookup('aaa_8'))
+        self.assertEqual(FOUND, codecs.lookup('AAA-8'))
+        self.assertEqual(FOUND, codecs.lookup('AAA---8'))
+        self.assertEqual(FOUND, codecs.lookup('AAA   8'))
+        self.assertEqual(FOUND, codecs.lookup('aaa\xe9\u20ac-8'))
+        self.assertEqual(NOT_FOUND, codecs.lookup('AAA.8'))
+        self.assertEqual(NOT_FOUND, codecs.lookup('AAA...8'))
+        self.assertEqual(NOT_FOUND, codecs.lookup('BBB-8'))
+        self.assertEqual(NOT_FOUND, codecs.lookup('BBB.8'))
+        self.assertEqual(NOT_FOUND, codecs.lookup('a\xe9\u20ac-8'))
+        self.addCleanup(codecs.unregister, search_function)
```

Comment:
It seems like you misunderstood the purpose of this function. It registers a function which is always called when a test complete, even if the test fails (with an exception). If you register the cleanup function only once all tests passed, your cleanup function will not be called if tests fail.

You have to call addCleanup just after your codecs.register() call. You can even call it **before** (the line before) since it's safe to unregister a codec search function which is not registered yet.

Reply:
Oh, thanks, victor, you are right. I will update it soon.